### PR TITLE
{S,E}QL: Fix optimization of `NotEquals` in conjunctions

### LINKED ISSUE #65322: SQL query not properly applying some operators
**Elasticsearch version** (`bin/elasticsearch --version`): 7.9.1

**Plugins installed**: []

**JVM version** (`java -version`):

**OS version** (`uname -a` if on a Unix-like system):

**Description of the problem including expected versus actual behavior**:
SQL query is properly translating all the operators when the query combine several conditions

**Steps to reproduce**:

It can be verified by using: https://demo.elastic.co/app/dev_tools#/console?load_from=https:%2F%2Fs3.us-east-2.amazonaws.com%2Fdemo.elastic.co%2Fquery_12.json&_g=()

1) Base query, shows first 3 row:
```
POST _sql?format=csv
{
"query":"SELECT FlightNum FROM flights ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
00882F6
009NIGR
00CGX81
```
2) Add `where` clause to unselect first row:
```
POST _sql?format=csv
{
"query":"SELECT FlightNum FROM flights WHERE FlightNum <> '00882F6' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
009NIGR
00CGX81
00DRRXR
```

Notice `00882F6` is not listed

3) Add `where` clause which does not change result set:
```
POST _sql?format=csv
{
"query":"SELECT FlightNum FROM flights WHERE timestamp >= '2010-06-18T22:58:43.000Z' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
00882F6
009NIGR
00CGX81
```

4) Combine both conditions yields a **wrong result**:
```
POST _sql?format=csv
{
"query":"SELECT FlightNum FROM flights WHERE timestamp >= '2010-06-18T22:58:43.000Z' AND FlightNum <> '00882F6' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
00882F6
009NIGR
00CGX81
```

Notice `00882F6` is there,  but should not be listed

Looking at the output of the `translate` endpoint, the query doesn't include any MUST_NOT 

```
POST _sql/translate
{
"query":"SELECT FlightNum FROM flights WHERE timestamp >= '2010-06-18T22:58:43.000Z' AND FlightNum <> '00882F6' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
{
  "size" : 3,
  "query" : {
    "range" : {
      "timestamp" : {
        "from" : "2010-06-18T22:58:43.000Z",
        "to" : null,
        "include_lower" : true,
        "include_upper" : false,
        "time_zone" : "Z",
        "boost" : 1.0
      }
    }
  },
  "_source" : {
    "includes" : [
      "FlightNum"
    ],
    "excludes" : [ ]
  },
  "sort" : [
    {
      "FlightNum.keyword" : {
        "order" : "asc",
        "missing" : "_last",
        "unmapped_type" : "keyword"
      }
    }
  ]
}
```

Using `!=` doesn't work either... But `=` is properly translated:

```
POST _sql/translate
{
"query":"SELECT FlightNum FROM flights WHERE timestamp >= '2010-06-18T22:58:43.000Z' AND FlightNum = '00882F6' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
{
  "size" : 3,
  "query" : {
    "bool" : {
      "must" : [
        {
          "range" : {
            "timestamp" : {
              "from" : "2010-06-18T22:58:43.000Z",
              "to" : null,
              "include_lower" : true,
              "include_upper" : false,
              "time_zone" : "Z",
              "boost" : 1.0
            }
          }
        },
        {
          "term" : {
            "FlightNum.keyword" : {
              "value" : "00882F6",
              "boost" : 1.0
            }
          }
        }
      ],
      "adjust_pure_negative" : true,
      "boost" : 1.0
    }
  },
  "_source" : {
    "includes" : [
      "FlightNum"
    ],
    "excludes" : [ ]
  },
  "sort" : [
    {
      "FlightNum.keyword" : {
        "order" : "asc",
        "missing" : "_last",
        "unmapped_type" : "keyword"
      }
    }
  ]
}


```

**Provide logs (if relevant)**:

**Workaround**:

As a workaround, we can use the date not as a string in the `where`clause:

`WHERE timestamp >= CAST ('2010-06-18T22:58:43.000Z' AS DATETIME)`

That gets properly translated:

```
POST _sql/translate
{
"query":"SELECT FlightNum FROM flights WHERE  timestamp >= CAST ('2010-06-18T22:58:43.000Z' AS DATETIME) AND FlightNum != '00882F6' ORDER BY FlightNum LIMIT 3"
}

RESPONSE
************
{
  "size" : 3,
  "query" : {
    "bool" : {
      "must" : [
        {
          "range" : {
            "timestamp" : {
              "from" : "2010-06-18T22:58:43.000Z",
              "to" : null,
              "include_lower" : true,
              "include_upper" : false,
              "time_zone" : "Z",
              "format" : "strict_date_time",
              "boost" : 1.0
            }
          }
        },
        {
          "bool" : {
            "must_not" : [
              {
                "term" : {
                  "FlightNum.keyword" : {
                    "value" : "00882F6",
                    "boost" : 1.0
                  }
                }
              }
            ],
            "adjust_pure_negative" : true,
            "boost" : 1.0
          }
        }
      ],
      "adjust_pure_negative" : true,
      "boost" : 1.0
    }
  },
  "_source" : {
    "includes" : [
      "FlightNum"
    ],
    "excludes" : [ ]
  },
  "sort" : [
    {
      "FlightNum.keyword" : {
        "order" : "asc",
        "missing" : "_last",
        "unmapped_type" : "keyword"
      }
    }
  ]
}

```

### PR DESCRIPTION
**Pull request by palesz**
* Fix the `CombineBinaryComparisons` optimizer rule, so that semantic
equality taken into account during the optimization of `NotEquals`

Examples that previously removed the `NotEquals` expressions (leading
to incorrect results):

```
double >= 10 AND integer != 9
-->  double >= 10

keyword != '2021' AND datetime >= '2020-01-01T00:00:00'
--> datetime >= '2020-01-01T00:00:00'
```

With the fix, expressions like the above will not be touched.
`NotEquals` will only be eliminated from the `AND` expression if the
left side of the `NotEquals` `semanticEquals()` to the left side
of the other expressions within the conjunction (comparisons against
the same field/expression).

* Unit tests and integration tests

Close #65322

Versions impacted: v7.7.0+. It is a question how far do we want to merge back the bugfix.

Bug was merged into master on Jan 30, 2020: https://github.com/elastic/elasticsearch/commit/789724ac2cc726de603849b4eeb8194da7528bcc
The bug first appeared in v7.7.0: https://github.com/elastic/elasticsearch/commit/f1173aaa486a32c5145f10db2bdd131103c4c639

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 1]
```diff
@@ -2338,4 +2338,12 @@ public void testInInRangeValues() {
         PhysicalPlan p = testContext.optimizeAndPlan("SELECT long FROM test WHERE long IN (1, 2, 3, " + Long.MAX_VALUE + ", 5, 6, 7)");
         assertEquals(EsQueryExec.class, p.getClass());
     }
+    
+    public void testCombineBinaryComparisonsBugKeyword() {
```
Comment (bpintea) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:2342]: I'd add a simple numeric test in x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java to replicate and check the fix, as well as more accurately inspect the translation.

In that respect, not sure if this particular test would still be needed, but if you decide to keep it, I'd also add a comment on the condition that the value of `negatedText` fulfils as it's not obvious.
Also, the "Bug" in the title is not explanatory: either rename the test or add a reference.

**Review by bpintea — COMMENTED** [review]
The fix looks good, thanks, left just some small suggestions around testing.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java [hunk 2]
```diff
@@ -135,7 +137,11 @@ private static FieldAttribute getFieldAttribute() {
     }
 
     private static FieldAttribute getFieldAttribute(String name) {
-        return new FieldAttribute(EMPTY, name, new EsField(name + "f", INTEGER, emptyMap(), true));
+        return getFieldAttribute(name, INTEGER);
+    }
+
+    private static FieldAttribute getFieldAttribute(String name, DataType dataType) {
+        return new FieldAttribute(EMPTY, name, new EsField(name + "f", dataType, emptyMap(), true));
     }
```
Comment (costin) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java:145]: Use the method in TestUtils.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 1 above)
Comment (palesz) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: Thanks for the comment. Cleaned up the draft, added additional test cases that were discovered by randomized tests (will add randomized tests in separate PR), added an integration test for this particular bug.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java [hunk 3]
```diff
@@ -1015,7 +1021,32 @@ public void testRangesOverlappingNoLowerBoundary() {
...
+    
+    public void testBinaryComparisonAndOutOfRangeNotEqualsDifferentFields() {
+        FieldAttribute fDouble = getFieldAttribute("double", DOUBLE);
+        FieldAttribute fDouble2 = getFieldAttribute("double2", DOUBLE);
+        FieldAttribute fInteger = getFieldAttribute("int", INTEGER);
+        FieldAttribute fDatetime = getFieldAttribute("datetime", INTEGER);
+        FieldAttribute fKeyword = getFieldAttribute("keyword", KEYWORD);
+        FieldAttribute fKeyword2 = getFieldAttribute("keyword2", KEYWORD);
+        ZoneId zoneId = randomZone();
```
Comment (costin) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java]: Better naming - doubleOne, doubleTwo - oneDouble, anotherDouble, etc..

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java [hunk 4]
```diff
@@ -1015,7 +1021,32 @@ public void testRangesOverlappingNoLowerBoundary() {
...
+            // double > 10 AND integer != -10
+            new And(EMPTY, new GreaterThan(EMPTY, fDouble, L(10), zoneId), new NotEquals(EMPTY, fInteger, L(-10), zoneId)),
+            // keyword > '5' AND keyword2 != '48'
+            new And(EMPTY, new GreaterThan(EMPTY, fKeyword, L("5"), zoneId), new NotEquals(EMPTY, fKeyword2, L("48"), zoneId)),
+            // keyword != '2021' AND datetime <= '2020-12-04T17:48:22.954240Z'
+            new And(EMPTY, new NotEquals(EMPTY, fKeyword, L("2021"), zoneId), 
+                new LessThanOrEqual(EMPTY, fDatetime, L("2020-12-04T17:48:22.954240Z"), zoneId)),
+            // double > 10.1 AND double2 != -10.1
+            new And(EMPTY, new GreaterThan(EMPTY, fDouble, L(10.1d), zoneId), new NotEquals(EMPTY, fDouble2, L(-10.1d), zoneId))
+        )) {
```
Comment (costin) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java:1045]: Extract the list outside the for.
Also use the methods in TestUtils to increase readability.

File: x-pack/plugin/sql/qa/server/src/main/resources/filter.sql-spec [hunk 5]
```diff
@@ -22,6 +22,10 @@ whereFieldAndComparison
 // tag::whereFieldAndComparison
 SELECT last_name l FROM "test_emp" WHERE emp_no > 10000 AND emp_no < 10005 ORDER BY emp_no LIMIT 5;
 // end::whereFieldAndComparison
+whereGreaterThanAndNotEqualityOnDifferentFields
+// tag::whereGreaterThanAndNotEqualityOnDifferentFields
+SELECT last_name l FROM "test_emp" WHERE salary >= 50000 AND emp_no != 10002 ORDER BY emp_no LIMIT 5;
+// end::whereGreaterThanAndNotEqualityOnDifferentFields
```
Comment (costin) [inline x-pack/plugin/sql/qa/server/src/main/resources/filter.sql-spec:28]: tag is used for documentation purposes. Unless you consume it into the docs, there's no point to add it.

**Review by costin — CHANGES_REQUESTED** [review]
Minor clean-up in testing.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java [hunk 6]
```diff
@@ -140,7 +143,11 @@ private static FieldAttribute getFieldAttribute() {
     }
 
     private static FieldAttribute getFieldAttribute(String name) {
-        return new FieldAttribute(EMPTY, name, new EsField(name + "f", INTEGER, emptyMap(), true));
+        return getFieldAttribute(name, INTEGER);
+    }
+
+    private static FieldAttribute getFieldAttribute(String name, DataType dataType) {
```
Comment (costin) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java:149]: Use TestUtils.fieldAttribute()

**Review by costin — APPROVED** [review]
The test can be improved further.

**Review by bpintea — APPROVED** [review]
LGTM, nice tests.

**Pull request merged**